crates/sage-cloudpath/src/lib.rs

```diff
@@ -263,9 +273,18 @@ pub enum Error {
 impl std::fmt::Display for Error {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            Error::InvalidUri => f.write_str("invalid URI"),
-            Error::S3(x) => x.fmt(f),
-            Error::IO(x) => x.fmt(f),
+            Error::InvalidUri => write!(f, "Invalid URI"),
+            Error::S3(x) => {
+                write!(f, "S3 error: ")?;
+                match x {
+                    // display a more informative error message than simply `unhandled error`
+                    aws_sdk_s3::Error::Unhandled(unhandled) => {
+                        writeln!(f, "{}", std::error::Error::source(unhandled).unwrap())
```

Comment:
Does this branch have an effect on the output?

I still see "S3 error: unhandled error"

Reply:
Weird. For me on my machine without any AWS config, I get

Before (f55a9e5)
```
; cargo run -- s3://my-bucket/parameters.json
   Compiling sage-cloudpath v0.11.0 (/Users/elendol/projects/sage/crates/sage-cloudpath)
   Compiling sage-core v0.12.0 (/Users/elendol/projects/sage/crates/sage)
   Compiling sage-cli v0.12.0 (/Users/elendol/projects/sage/crates/sage-cli)
    Finished dev [unoptimized + debuginfo] target(s) in 3.27s
     Running `target/debug/sage 's3://my-bucket/parameters.json'`
Error: unhandled error
```

After (16138c8)

```
; cargo run -- s3://my-bucket/parameters.json
   Compiling sage-cloudpath v0.11.0 (/Users/elendol/projects/sage/crates/sage-cloudpath)
   Compiling sage-core v0.12.0 (/Users/elendol/projects/sage/crates/sage)
   Compiling sage-cli v0.12.0 (/Users/elendol/projects/sage/crates/sage-cli)
    Finished dev [unoptimized + debuginfo] target(s) in 2.87s
     Running `target/debug/sage 's3://my-bucket/parameters.json'`
Error: Failed to read parameters from `s3://my-bucket/parameters.json`

Caused by:
    S3 error: failed to construct request
    
```

Similar if I try to pass a S3 uri for the database.

For the mzML files it goes from (f55a9e5):

```
; cargo run -- tests/config.json  s3://my-bucket/ms.mzML
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/sage tests/config.json 's3://my-bucket/ms.mzML'`
[2023-05-08T19:52:31Z INFO  sage] generated 1845 fragments in 9ms
[2023-05-08T19:52:31Z INFO  sage] processing files 0 .. 1 
[2023-05-08T19:52:34Z ERROR sage] error while processing s3://my-bucket/ms.mzML: unhandled error
[2023-05-08T19:52:34Z INFO  sage]  - file IO:     2194 ms
[2023-05-08T19:52:34Z INFO  sage]  - search:         0 ms (0 spectra)
[2023-05-08T19:52:34Z INFO  sage_core::ml::retention_alignment] aligning file #0: y = NaNx + NaN
[2023-05-08T19:52:34Z INFO  sage_core::ml::retention_alignment] aligned retention times across 1 files
[2023-05-08T19:52:34Z INFO  sage_core::ml::retention_model] - fit retention time model, rsq = NaN
[2023-05-08T19:52:34Z WARN  sage_core::ml::gauss] Finding solution to linear system failed: left side of matrix [0,0] = NaN
[2023-05-08T19:52:34Z WARN  sage] linear model fitting failed, falling back to heuristic discriminant score
[2023-05-08T19:52:34Z INFO  sage] discovered 0 target peptide-spectrum matches at 1% FDR
```

To (16138c8):

```
; cargo run -- tests/config.json  s3://my-bucket/ms.mzML
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/sage tests/config.json 's3://my-bucket/ms.mzML'`
[2023-05-08T19:44:51Z INFO  sage] generated 1845 fragments in 9ms
[2023-05-08T19:44:51Z INFO  sage] processing files 0 .. 1 
[2023-05-08T19:44:53Z ERROR sage] error while processing s3://my-bucket/ms.mzML: S3 error: failed to construct request

[2023-05-08T19:44:53Z INFO  sage]  - file IO:     2198 ms
[2023-05-08T19:44:53Z INFO  sage]  - search:         0 ms (0 spectra)
[2023-05-08T19:44:53Z INFO  sage_core::ml::retention_alignment] aligning file #0: y = NaNx + NaN
[2023-05-08T19:44:53Z INFO  sage_core::ml::retention_alignment] aligned retention times across 1 files
[2023-05-08T19:44:53Z INFO  sage_core::ml::retention_model] - fit retention time model, rsq = NaN
[2023-05-08T19:44:53Z WARN  sage_core::ml::gauss] Finding solution to linear system failed: left side of matrix [0,0] = NaN
[2023-05-08T19:44:53Z WARN  sage] linear model fitting failed, falling back to heuristic discriminant score
[2023-05-08T19:44:53Z INFO  sage] discovered 0 target peptide-spectrum matches at 1% FDR
```
Here the error log already mentioned the problematic file, but we now have the underlying error in case of `unhandled error` (it might be possible that the underlying error in some cases might still be `unhandled error`, I will dig more into the error types for the AWS SDK).

There is an unnecessary `\n` I am going to remove though. 

Also I believe sage should fail in case of missing input files.

Reply:
This is great, nice find!

I think this actually eliminates the need for the above empty key check. Removing that check yields the following error:

```
S3 error: unhandled error: failed to construct request: key was missing: cannot be empty or unset (Unhandled { source: ConstructionFailure(ConstructionFailure { source: BuildError { kind: MissingField { field: "key", details: "cannot be empty or unset" } } }) })
```

This seems more informative than "Invalid URI"

Reply:
Actually, just reading the source from the aws ErrorContext type, I think copying it and modifying it can produce even cleaner error message: 
```rust
write!(f, "S3 error - ")?;
                match x {
                    // display a more informative error message than simply `unhandled error`
                    aws_sdk_s3::Error::Unhandled(unhandled) => {
                        let mut stack = std::error::Error::source(&unhandled);
                        while let Some(source) = stack {
                            write!(f, "{}. ", source)?;
                            stack = std::error::Error::source(source);
                        }
                        Ok(())
                    }
                    // other error kinds should already be more informative
                    _ => x.fmt(f),
                }
```

Reply:
And yes, we probably should fail on missing files & failed mzML parsing

Reply:
I just wish we didn't have to manually open the unhandled error.